libs/postgres_ffi/src/xlog_utils.rs

```diff
@@ -224,42 +264,103 @@ fn find_end_of_wal_segment(
 
             // read the rest of the record, or as much as fits on this page.
             let n = min(contlen, pageleft);
-            // fill rec_hdr (header up to (but not including) xl_crc field)
+            trace!(
+                "offs=0x{:x}, record continuation, pageleft={}, contlen={}",
+                offs,
+                pageleft,
+                contlen
+            );
+            // fill rec_hdr header up to (but not including) xl_crc field
+            trace!(
+                "  rec_offs={}, XLOG_RECORD_CRC_OFFS={}, XLOG_SIZE_OF_XLOG_RECORD={}",
+                rec_offs,
+                XLOG_RECORD_CRC_OFFS,
+                XLOG_SIZE_OF_XLOG_RECORD
+            );
             if rec_offs < XLOG_RECORD_CRC_OFFS {
                 let len = min(XLOG_RECORD_CRC_OFFS - rec_offs, n);
+                trace!(
+                    "  reading rec_hdr[{}..{}] <-- [0x{:x}; 0x{:x})",
+                    rec_offs,
+                    rec_offs + len,
+                    page_offs,
+                    page_offs + len
+                );
                 rec_hdr[rec_offs..rec_offs + len].copy_from_slice(&buf[page_offs..page_offs + len]);
             }
+            // TODO: handle `rec_offs + n < XLOG_SIZE_OF_XLOG_RECORD`: the header is not completely read.
```

Comment:
Sorry, I do not understand the problem.
Yes, record head can be cut at any 8-byte word. Only 8-byte alignment is enforced.
So it can happen that record header starts at one page, but CRC is at another page.

But this accumulation of page header in  `rec_hdr` is used to address this case. CRC field can not be splitted between two different pages, because of 8-byte alignment. So the following check `rec_offs <= XLOG_RECORD_CRC_OFFS && rec_offs + n >= XLOG_SIZE_OF_XLOG_RECORD` should catch moment when we have CRC field in the buffer.

Reply:
It will be accumulated in `rec_hdr`, but it will also update `crc` with a piece of header in the `else`-branch below.

Ah, I see, it does not matter as we reset `crc` back to zero in the `if` branch once we have read the header.